cipher-suite/src/main/kotlin/net/corda/v5/cipher/suite/merkle/MerkleTreeProvider.kt

```diff
@@ -0,0 +1,44 @@
+package net.corda.v5.cipher.suite.merkle
+
+import net.corda.v5.crypto.DigestAlgorithmName
+import net.corda.v5.base.annotations.DoNotImplement
+import net.corda.v5.base.annotations.Suspendable
+import net.corda.v5.crypto.extensions.merkle.MerkleTreeHashDigestProvider
+import net.corda.v5.crypto.merkle.MerkleTree
+import net.corda.v5.crypto.merkle.MerkleTreeHashDigest
+
+/**
+ * [MerkleTreeProvider] creates [MerkleTree]s and [MerkleTreeHashDigest]s.
+ */
+@DoNotImplement
+interface MerkleTreeProvider {
+    /**
+     * Creates a [MerkleTree]
+     *
+     * @param leaves The leaves of the tree.
+     * @param digestProvider Merkle Tree Hash digest provider used to construct the tree's node and leaf hashes.
+     *
+     * @return A new [MerkleTree] instance.
+     */
+    @Suspendable
+    fun createTree(
+        leaves: List<ByteArray>,
+        digestProvider: MerkleTreeHashDigestProvider
+    ) : MerkleTree
```

Comment:
What's the benefit of having `MerkleTreeProvider` as well as `MerkleTreeFactory`? We can create `MerkleTreeProviderImpl` and proxy call into it from `MerkleTreeFactoryImpl`, but it will still return `MerkleTree` which relies on the new marker interface `MerkleTreeHashDigest` rather than `MerkleTreeHashDigestProvider`.

Reply:
Alright, so to stick with that pattern, we should create `MerkleTreeProviderImpl` in corda-runtime-os which will basically do what `MerkleTreeFactoryImpl` did? And have `MerkleTreeFactoryImpl` just delegate to the provider impl?

Reply:
Yep